spec/models/package_spec.rb

```diff
@@ -169,20 +169,269 @@
+      }.to change(PackagesLocation, :count).by(0)
+    end
+  end
+
+  describe '#move_full_quantity' do
+    let!(:package) { create :package }
+    let!(:location) { create :location }
+    let!(:order) { create :order, state: "submitted"}
+    let!(:orders_package) { create :orders_package, package: package, state: 'designated', order: order, quantity: 10 }
+    let!(:packages_location) { create :packages_location, package: package, reference_to_orders_package: orders_package.id}
+
+    context 'if no packages_location record exist with provided location_id' do
+      it 'updates quantity of packages_location record with referenced orders package quantity' do
+        package.move_full_quantity(location.id, orders_package.id)
+        expect(packages_location.reload.quantity).to eq orders_package.quantity
+      end
+
+      it 'updates location_id of packages_location record with provided location_id' do
+        package.move_full_quantity(location.id, orders_package.id)
+        expect(packages_location.reload.location).to eq location
+      end
+
+      it 'clears reference_to_orders_package' do
+        package.move_full_quantity(location.id, orders_package.id)
+        expect(packages_location.reload.reference_to_orders_package).to be_nil
+      end
+    end
+
+    context 'if packages_location record already exist with provided location_id' do
+      let!(:packages_location_1) { create :packages_location, package: package, reference_to_orders_package: orders_package.id, location: location}
+
+      it 'adds up orders_package quantity and packages_location quantity and updates packages_location quantity with new quantity' do
+        new_quantity = packages_location_1.quantity + orders_package.quantity
+        package.move_full_quantity(location.id, orders_package.id)
+        expect(packages_location_1.reload.quantity).to eq new_quantity
+      end
+
+      it 'destroys referenced packages_location' do
+        expect{
+          package.move_full_quantity(location.id, orders_package.id)
+        }.to change(PackagesLocation, :count).by(-1)
+      end
+
+      it 'clears reference_to_orders_package' do
+        package.move_full_quantity(location.id, orders_package.id)
+        expect(packages_location_1.reload.reference_to_orders_package).to be_nil
+      end
+    end
+  end
+
+  describe '#move_partial_quantity' do
+    let!(:package) { create :package }
+    let!(:location) { create :location }
+    let!(:location_1) { create :location }
+    let!(:packages_location) { create :packages_location, quantity: 12, package: package, location: location_1 }
+
+    context 'moving some qty to location for which associated packages_location do not exist' do
+      it 'subtract quantity to move from packages_location record(current location)' do
+        quantity_to_move = 5
+        package_qty_changes = [{packages_location_id: packages_location.id, package_id: package.id, new_qty: quantity_to_move}].to_json
```

Comment:
see previous note about moving JSON.parse... you won't need to do this step then